client/src/extension.ts

```diff
@@ -5,46 +5,145 @@
+
 let client: LanguageClient;
 
-async function isEsyProject() {
+async function getCommandForWorkspace(context): Promise<{ command: string, args: string[] }> {
+
   let root = workspace.rootPath;
-  if (await exists(path.join(root, "package.json"))) {
-    return true;
-  } else if (await exists(path.join(root, "esy.json"))) {
-    return true;
-  } else {
-    return false;
-  }
-}
+  let esyStatus = await getEsyStatus(root);
+  if (esyStatus.isProject) {
+
+    // All npm and opam projects are valid esy projects too! Picking
+    // the right package manager is important - we don't want to run
+    // `esy` for a user who never intended to. Example: bsb/npm
+    // users. Similarly, opam users wouldn't want prompts to run
+    // `esy`. Why is prompting `esy i` even necessary in the first
+    // place? `esy ocamlmerlin-lsp` needs projects to install/solve deps
+
+    let manifestFile = esyStatus.rootPackageConfigPath;
+
+    // Looking for manifest file. If it is not a json file, it's an opam project
+    if (/\.json$/.test(manifestFile)) {
+
+      let manifest;
+      try {
+        manifest = fs.readFileSync(manifestFile).toString();
+      } catch (e) {
+        return Promise.reject(e);
+      }
+
+      try {
+        manifest = JSON.parse(manifest);
+      } catch (e) {
+        return Promise.reject(e);
+      }
+
+      if (/esy\.json$/.test(manifestFile) || !!manifest.esy) {
+        // Esy is indeed being used to manage this project
+        log('Ensuring toolchain is up-to-date by running `esy`');
+      } else {
+        // Esy may not have been used. Let's drop and esy.json with the an appropriate compiler, reason and merlin-lsp
+        // Note that this check (npm vs esy) needs to be done only once. Since all npm projects will eventually end up having an esy.json with the toolchain
+        let ocaml: string;
+        if (manifest.dependencies && manifest.dependencies['bs-platform'] && semver.lt(manifest.dependencies['bs-platform'], '6.x')) {
+          ocaml = '4.2.x';
+        } else {
+          ocaml = '4.6.x';
+        }
+        let reason = "*";
+        let merlinLsp = "ocaml/merlin:merlin-lsp.opam#f030d5da7a"
+
+        // Creating esy.json
+        fs.writeFileSync(path.join(root, 'esy.json'), JSON.stringify({ dependencies: { ocaml, "@esy-ocaml/reason": reason, "merlin-lsp": merlinLsp } }));
+
+        // Running esy i && esy b
+        log('Creating esy.json and setting up toolchain');
+      }
+      window.withProgress({
+        location: ProgressLocation.Notification,
+        title: "Setting up toolchain...",
+        cancellable: true
+      }, (progress, token) => {
+        token.onCancellationRequested(() => {
+          console.log("User canceled the long running operation");
+        });
+
+        progress.report({ increment: 0 });
+
+        setTimeout(() => {
+          progress.report({ increment: 10 });
+        }, 1000);
+
+
+        return run('esy', { cwd: root })

```

Comment:
fwiw `esy status` output shows if all deps are built or not, we can skip running esy in this case.